cmd/repo-query/repo-query.go

```diff
@@ -0,0 +1,353 @@
+
+		if strings.HasPrefix(line, "event: ") {
+			chunk := line
+			for {
+				line, err := reader.ReadString('\n')
+				if err != nil {
+					return err
+				}
+
+				if line == "\n" {
+					break
+				}
+
+				chunk += line
+			}
+
+			err := processIndexChunk(chunk)
+			if err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func processIndexChunk(chunk string) error {
+	chunkLines := strings.Split(chunk, "\n")
+	eventLine := chunkLines[0]
+	event := strings.Split(eventLine, ": ")[1]
+
+	switch event {
+	case "FETCH_REPO":
+		fmt.Println("Fetching Repository from GitHub...")
+	case "EMBED_REPO":
+		fmt.Println("Embedding Repository...")
+	case "SAVE_EMBEDDINGS":
+		fmt.Println("Saving the embeddings to our database...")
+	case "ERROR":
+		fmt.Println("There was an error while indexing this repository.")
+		return errors.New("error while indexing repository")
+	case "DONE":
+		fmt.Println("Indexing Complete. You can now ask questions about this repository! 🎉")
+	default:
+		break
+	}
+
+	return nil
+}
+
+type queryPostRequest struct {
+	Query      string `json:"query"`
+	Repository struct {
+		Owner  string `json:"owner"`
+		Name   string `json:"name"`
+		Branch string `json:"branch"`
+	} `json:"repository"`
+}
+
+func askQuestion(question string, owner string, repo string, branch string) error {
+	queryPostReq := &queryPostRequest{
+		Query: question,
+		Repository: struct {
+			Owner  string `json:"owner"`
+			Name   string `json:"name"`
+			Branch string `json:"branch"`
+		}{
+			Owner:  owner,
+			Name:   repo,
+			Branch: branch,
+		},
+	}
+
+	queryPostJSON, err := json.Marshal(queryPostReq)
+	if err != nil {
+		return err
+	}
+
+	responseBody := bytes.NewBuffer(queryPostJSON)
+	resp, err := http.Post(fmt.Sprintf("%s/query", repoQueryURL), "application/json", responseBody)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return err
+		}
+
+		fmt.Printf("An error occurred: %v\n", string(body))
+		return errors.New("error while asking question")
+	}
+
+	//  listen for SSEs and send data,event pairs to processChatChunk
+	reader := bufio.NewReader(resp.Body)
+	for {
+		line, err := reader.ReadString('\n')
+		// if err is eof then break
+		if err == io.EOF {
+			return nil
+		} else if err != nil {
+			return err
+		}
+
+		if line == "\n" {
+			continue
+		}
+
+		if strings.HasPrefix(line, "event: ") {
+			chunk := line
+			for {
+				line, err := reader.ReadString('\n')
+				if err != nil {
+					return err
+				}
+
+				if line == "\n" {
+					break
+				}
+
+				chunk += line
+			}
+
+			err := processChatChunk(chunk)
+			if err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func processChatChunk(chunk string) error {
+	chunkLines := strings.Split(chunk, "\n")
+	eventLine := chunkLines[0]
+	dataLine := chunkLines[1]
+
+	event := strings.Split(eventLine, ": ")[1]
+	var data interface{} // the data can be a string or a JSON object
```

Comment:
> the data can be a string or a JSON object

Hmmm that's unfortunate. Can this returned data be normalized somehow? Might be worth ensuring the server side is returning normalized data.

What you have is fine, but attempting to cast `data.(map[string]interface{})` will panic if this is not a `map[string]interface{}` type. I.e., if this is indeed a string for some reason, this will break and panic mid flight.

It looks like the only case where `data` is a string is during `DONE`. Can that be normalized on the server side? Might need @Anush008's comments too.

Reply:
Yeah, he said he'll normalise it, but what you said is what we have currently (string on `DONE` and json for other events)